File: Group 3 - Dynamic Programming/Q13 - Minimal Grid Path.cpp
// https://cses.fi/problemset/task/3359/
// direct c++ translation
// check Python code for explanation

#include <bits/stdc++.h>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<string> grid(n);
    for (int i = 0; i < n; i++) {
        cin >> grid[i];
    }

    vector<vector<char>> vis(n, vector<char>(n, 0));

    vector<int> q1, q2;
    q1.push_back(0);
    q2.push_back(0);

    string res;
    res.reserve(2 * n);

    for (int le = 1; le < 2 * n; le++) {
        // find minimum character in current layer
        char min_char = 'z';
        for (int i = 0; i < (int)q1.size(); i++) {
            min_char = min(min_char, grid[q1[i]][q2[i]]);
        }
        res.push_back(min_char);

        // keep only positions achieving min_char
        vector<int> nq1, nq2;
        for (int i = 0; i < (int)q1.size(); i++) {
            if (grid[q1[i]][q2[i]] == min_char) {
                nq1.push_back(q1[i]);
                nq2.push_back(q2[i]);
            }
        }
        q1.swap(nq1);
        q2.swap(nq2);

        // expand BFS frontier
        vector<int> nxt1, nxt2;
        for (int i = 0; i < (int)q1.size(); i++) {
            int r = q1[i], c = q2[i];

            if (r + 1 < n && !vis[r + 1][c]) {
                vis[r + 1][c] = 1;
                nxt1.push_back(r + 1);
                nxt2.push_back(c);
            }
            if (c + 1 < n && !vis[r][c + 1]) {
                vis[r][c + 1] = 1;
                nxt1.push_back(r);
                nxt2.push_back(c + 1);
            }
        }

        q1.swap(nxt1);
        q2.swap(nxt2);
    }

    cout << res << '\n';
    return 0;
}

File: Group 3 - Dynamic Programming/Q13 - Minimal Grid Path.py
# TLE IN PYTHON
#
# https://cses.fi/problemset/task/3359
# Greedy + BFS, 10-12p
#
# Since we are looking for lex min, always prioritize early characters -> greedy
# If there is a tie, we maintain all indices that create optimal strings
#
# TC: O(n^2)

n = int(input())
grid = [input() for _ in range(n)]

vis = [[False] * n for _ in range(n)]
res = []
q1 = [0]  # parallel list to avoid tuples
q2 = [0]
for le in range(1, 2 * n):  # layer by layer BFS
    # only keep optimal (and tied) characters
    min_char = min(grid[r][c] for r, c in zip(q1, q2))
    res.append(min_char)
    keep_idx = [i for i in range(len(q1)) if grid[q1[i]][q2[i]] == min_char]
    q1 = [q1[i] for i in keep_idx]
    q2 = [q2[i] for i in keep_idx]

    nxt1 = []
    nxt2 = []
    for r, c in zip(q1, q2):  # consider next characters
        if r + 1 < n and not vis[r + 1][c]:
            vis[r + 1][c] = True
            nxt1.append(r + 1)
            nxt2.append(c)
        if c + 1 < n and not vis[r][c + 1]:
            vis[r][c + 1] = True
            nxt1.append(r)
            nxt2.append(c + 1)

    q1 = nxt1
    q2 = nxt2

print("".join(res))

# # scrapped idea, this is overkill as it solves for all possible starting points
# Grid DP + reconstruction, 12-15p
#
# Consider the strings formed by starting at (r,c) and going to the end
# We only need to track the relative lexicographical order of strings of the same length
#
# dp[r][c] = relative lex order of string starting at (r,c) compared to equal length strings
# We do the transitions from small to large length (i.e. 1 diagonal at a time)
# When doing transitions, prioritize the early letter, tiebreak by rest of the string
#
# TC: O(n^2 * log(n))
